### DISCUSSION TITLE
A friendlier alternative for `get_action()`

### DISCUSSION THREAD

**Discussion by amercader** [ckan/ckan#6890] [General]
Can we make this function, one of the most important for plugins a little bit friendlier?

```python
result = get_action("package_search")(
   {"user": g.user, "ignore_auth": True}, {"q": "salmon", "sort": metadata_modified desc"}
)
```
Our tests have used the [`call_action()`](https://github.com/ckan/ckan/blob/91c9366cf8df0dc4e12a9a043756eda5e40534b0/ckan/tests/helpers.py#L85) helper for many years, and it would be nice to finally realise @seanh [old vision](https://github.com/ckan/ckan/blob/91c9366cf8df0dc4e12a9a043756eda5e40534b0/ckan/tests/helpers.py#L108) of making it widely available (I would not deprecate `get_action()`, I think it's really useful some times)

Things I dislike about `get_action()`:

* My main issue with it is that it's confusing for newcomers.Our API docs list each action params, which makes sense for use of the HTTP API interface but when you need to call it from a Python plugin you need to learn about the "context" and "data_dict" objects. For people relatively new to Python the first bit where a callable is returned and then you call it with the two dicts takes time to wrap your head about it and is error prone.
* We are putting the context, a thing we generally want to avoid overusing and sharing across actions, right in front of the function call. Even if you don't care about it you need to pass an empty dict or None to pass your action parameters:
     ```python
     result = get_action("package_search")({}, {"q": "salmon", "sort": metadata_modified desc"})
     ```
* IMO it's just awkward :)

## Alternatives

I suggest we implement a `call_action()` method that in turns calls `get_action()` to keep all the existing nice things that `get_action` does (pre-populate the context, chain actions, etc).

Any keyword arguments passed to `call_action()` are wrapped in the `data_dict` passed to `get_action`:

```python
result = call_action("package_search", q="salmon", sort="metadata_modified desc")

result = call_action("package_update", **big_dataset_dict)

```
In most cases that's all that users need, as `get_action` will populate `model` (for whatever this is used) and `user` in the `context`. If you need to pass a context for some reason you could pass it explicitly with a `context` param:

```python
result = call_action("package_search", q="salmon", sort="metadata_modified desc", context={"ignore_auth": True})
```

What do people think? What did I miss?
Can we go even further with the context stuff to provide a nicer interface? eg dedicated `ignore_auth` or `user_auth` parameters that populate the context under the hood

**Comment by wardi** [comment]
including context at the same level as the `data_dict` parameters makes your example above dangerous:
```python
result = call_action("package_update", **big_dataset_dict)
```
If `big_dataset_dict` comes from the user, then a user will be able to pass values to the context e.g. to bypass authorization checks or cause software failures (there is no validation done on context because it's expected to come from a trusted source).

Some alternatives:

### 1. position-only parameter for context
```python
def call_action(action_name, context=None, /, **data_dict):
```
would look the same as your suggestion above except when passing a context:

```python
result = call_action("package_search", {"ignore_auth": True}, q="salmon", sort="metadata_modified desc")
```

### 2. different functions for passing context or overriding common paramaters
```python
def call_action_as_user(action_name, user, /, **data_dict):
    ...
def call_action_with_context(action_name, context, /, **data_dict):
    ...
```
Note that it's still probably a good idea in this case to use positional-only parameters in case `data_dict` contains `action_name`, `user` or `context` keys.

### 3. ckanapi `action` attribute-style

We could follow the ckanapi action shortcut style to make calling actions look even more like normal function calls.

```python
result = action.package_search(q="salmon", sort="metadata_modified desc")
```
 For cases where we need a context it could be completely separated something like:

```python
result = action_with_context(ignore_auth=True).package_search(q="salmon", sort="metadata_modified desc")
```
### 4. ckanapi `call_action`-style

ckanapi's "lower level" `call_action` methods elides any parameter magic, and separates files into their own parameter acknowledging the limitation of our action API in having any nested parameters when files are passed:

```python
call_action(action, data_dict=None, context=None, files=None):
```
This is geared to cases where `data_dict` is assembled within an action over many lines of code (*very* common in our code base) but not as pretty as the other options for simple calls.

```python
result = call_action("package_search", {"q":"salmon", "sort":"metadata_modified desc"}, {"ignore_auth": True})
```
Unfortunately this is the opposite parameter order as seen in the actual functions. I chose this order because it's more common to want to pass data_dict parameters than context ones.

**Comment by smotornyuk** [comment]
I like the first option, but the 3rd looks as most beginner-friendly, probably. Because the idea of context is not really clear unless you've spent some time messing with the API.

But there is a problem with implicit/optional context. Let's imagine the action `X` that calls another action `Y`. In the current form, whoever created action `X`, that person has likely done something like 
```python
def X(context, data_dict):
    return tk.get_action("Y")(context.copy(), data_dict)
```

If the developer knows how context works, you'll likely see something similar. If the dev doesn't know how context works, he'll just do such a thing to stay on the safe side(or because he doesn't know that context can be omitted).
But without explicit context, aka `action.Y(a=1)` or `call_action('Y', a=1)`, context will likely get lost. Devs who just blindly passed context earlier, will not do this anymore. And `Y` will be called with the default context, i.e `{"user": toolkit.g.user}`. This can be a problem when:

* you are intentionally using `site_user` or `ignore_auth` when calling `X`. When I am doing such a thing, I want all underlying actions to be called with the same context. 
* you are calling actions from CLI. Even if you send a context with a `user` to `X` action, the context will not be passed to the underlying `Y` and you'll get `RuntimeError`, because `tk.g.user` is not set.

In addition, even now there is a subtle problem with context. If an action calls a helper which relies on the current user, `tk.g.user` will be used, because helpers have no access to context(unless it's an explicit argument of the helper). I know that one should never do that because helpers are registered via the i-**TEMPLATE**-helpers interface. But everyone does it sometimes. So it would be nice to have an option to get `context['user']`, not the `tk.g.user` not only from the action but from everywhere. 

*Of course, it's not a problem, if the dev knows how CKAN internals, but I don't think that we can rely on it:)


So... if we change the context a bit, the problem may be solved and `context` will not be used as a part of the action call. For example, `contextvars` can be used here. With some hypothetical `set_context` and `get_context` functions, we can do the following:

```

def action(data_dict):
    # can be called from anywhere and return current value of the user-contextvar.
    user = get_context("user") 
   
    # changes context only inside this block
    with set_context(user=ANOTHER_USER):
        # assert user != get_context("user")
        call_action('another_action')

    # context is restored here
    assert user == get_context("user")

```

**Reply by wardi** [reply]
hmm. adding more magic to our situation with contextvars feels like a bad idea.
The context parameter is used a bunch of different ways that probably should be separated:
- auth information (user, ignore_auth)
- special internal flags only applicable to the current call
- special return values (package, resource objects etc.) shared in lots of scary ways

**Reply by wardi** [reply]
We do already have some thread-locals like contextvars for things like g.user as you mentioned, and the current language setting for `gettext`. Should we stop setting context['user'] and context['ignore_auth'] and instead use a context manager like your example and values in `g`? That would propagate auth information without parameter passing and keep it all in the same place.

**Reply by smotornyuk** [reply]
I'd like to do it in such a way and get rid of the current action's context. But I don't want to rely on the `g` object and such thread-local proxies, because most of them came from the Flask and they are mainly bound to the request context. Currently, we have to mock up request context in the tests\CLI and in some edge cases, like the `url_for` helper. And it creates an additional layer of issues and complexity for people, who don't know enough about CKAN internals.

That's why I suggested `contextvars` as a way to unify the approach. It's not necessarily the only way, and we can use thread-local proxies like `g`. But in that case, we need to put some effort and try to completely eliminate RuntimeErrors from the CKAN. Devs shouldn't think about the Flask and shouldn't even know that it's used under the hood unless they are creating Blueprint or Middleware.

**Reply by wardi** [reply]
reading more about contextvars I think they are an improvement over our context dict use. They clean up after themselves and are easy to grep for.

**Comment by amercader** [comment]
Thanks @wardi.

I think option 1 would be perfect but if I'm not mistaken position-only parameters are only available on Python 3.8. IIRC we agreed to support Python 3.7 for the next CKAN version.

What about a version 2 without the position-only params? So we have functions with dedicated params for context and user and we just make sure to `pop()` any `context` key from the incoming keyword params. Unfortunately there is an action ([`api_token_create`](http://docs.ckan.org/en/latest/api/index.html#ckan.logic.action.create.api_token_create)) that expects a `user` param :upside_down_face: so if we want to put a safety check for that particular key we either special-case it or rename the param (which seems a reasonable API breakage to be honest).

I also like option 3. Maybe that even would allow in the future to properly type hint the data_dict params of each action so you could get auto-completion of action params?

**Reply by smotornyuk** [reply]
Yep, position-only params are not available in v3.7, but they can be replaced by an uglier version until we drop v3.7 support:
```
def call_action(*params, **data_dict):
  if len(params) == 1:
    action_name = params[0]
    context = None
  elif len(params) == 2:
    action_name, context = params:
  else:
    assert False
  ...

```
